apps/src/music/views/MusicView.jsx

```diff
@@ -100,9 +102,16 @@ class UnconnectedMusicView extends React.Component {
     // TODO: the 'beforeunload' callback is advised against as it is not guaranteed to fire on mobile browsers. However,
     // we need a way of reporting analytics when the user navigates away from the page. Check with Amplitude for the
     // correct approach.
-    window.addEventListener('beforeunload', () =>
-      this.analyticsReporter.endSession()
-    );
+    window.addEventListener('beforeunload', event => {
+      this.analyticsReporter.endSession();
+      // force a save before the page unloads, if there are unsaved changes.
+      // If we need to force a save, prevent navigation so we can save first.
+      if (this.musicBlocklyWorkspace.hasUnsavedChanges()) {
```

Comment:
As we add this to more labs, I'm wondering if this is something the ProjectManager (or whatever parent lab component we use) should set up itself, so we get this behavior automatically for all labs? This is definitely fine for now, just wanted to note it down

Reply:
Good question. I think once we have the lab container component we can look into something like that, but I'll leave it as-is for now.